File: testsuite/tests/ada/0008-inheritance/cpp/main.cpp
#include "gnatpolyglot_ada_arrays.h"
#include "test.h"

#include <iostream>
#include <string>
#include <vector>

class OtherChild : public test::Root {
public:
    OtherChild(const std::string &data) : test::Root(4, 2, this), _str(data) {}

    void p1() const override{
        std::cout << _str << "\n";
    }

    test::Rec f() const override {
       return test::Rec(21);
    }

    gnatpolyglot::ada::arrays::polyglot_array<int32_t>
    f_arr(const gnatpolyglot::ada::arrays::polyglot_array<int32_t> &a, int &i) const override {
        gnatpolyglot::ada::arrays::polyglot_array<int32_t> res(a.get_begin(), a.get_end());
        for (int i = res.get_begin(); i <= res.get_end(); i++) {
            res.set(i, a.get(i) + 2);
        }
        std::cout << "Returning from C++: ";
        for (int i = res.get_begin(); i <= res.get_end(); i++) {
            std::cout << res.get(i) << ", ";
        }
        std::cout << "\n";
        i += 2;
        return res;
    }

private:
    std::string _str;
};

class GrandChild : public test::Child {
public:
    GrandChild(const std::vector<int> &data) : test::Child(4, 2, 1, this), _vec(data) {}

    void p1() const override {
        for (int i : _vec) {
            std::cout << i << " ";
        }
        std::cout << "\n";
    }

    void p2(const test::Rec &rec, int32_t i) const override {
        std::cout << "P2 from C++ " << i << ": " << rec.get_i() << "\n";
    }

    test::Rec f() const override {
        return test::Rec(100);
    }

private:
    std::vector<int> _vec;
};

void test_array(test::Root &r, gnatpolyglot::ada::arrays::polyglot_array<int32_t> input) {
    int i = 0;
    gnatpolyglot::ada::arrays::polyglot_array<int32_t> res = test::f_arr_disp(r, input, i);
    std::cout << "Got in C++: ";
    for (int i = res.get_begin(); i <= res.get_end(); i++) {
        std::cout << res.get(i) << ", ";
    }
    std::cout << "\n" << "i = " << i << "\n";

}

int main() {
    test::Root root(1, 2);

    root.p1();

    test::Child child(3, 4, 5);

    child.p1();

    OtherChild other("c++ inherited");
    other.p1();

    test::Root &ref = other;
    ref.p1();

    GrandChild grand({3, 4, 5});
    grand.p1();
    test::Child &gref = grand;
    gref.p1();
    test::Root &ref2 = gref;
    ref2.p1();

    test::p_root(root);
    test::p_root(child);
    test::p_root(other);
    test::p_root(grand);

    test::Rec r1 = test::f_rec(root);
    std::cout << "r1 = " << r1.get_i() << "\n";
    test::Rec r2 = test::f_rec(ref);
    std::cout << "r2 = " << r2.get_i() << "\n";
    test::Rec r3 = test::f_rec(grand);
    std::cout << "r3 = " << r3.get_i() << "\n";

    test::p2_child(child, r1, 1);
    test::p2_child(gref, r2, 1);

    gnatpolyglot::ada::arrays::polyglot_array<int32_t> arr(1, 5);
    for (int i = arr.get_begin(); i <= arr.get_end(); i++) {
        arr.set(i, i);
    }
    test_array(root, arr);
    test_array(other, arr);
}

File: testsuite/tests/ada/0008-inheritance/java/Main.java
import com.adacore.libtest.test.Child;
import com.adacore.libtest.test.Rec;
import com.adacore.libtest.test.Root;
import com.adacore.libtest.test.TestPackage;
import com.adacore.gnatpolyglot.runtime.ada2java.IntegerArray;
import com.adacore.gnatpolyglot.runtime.IntegerRef;

import java.util.List;

public class Main {

    public static class OtherChild extends Root {

        private String str;

        public OtherChild(String str) {
            super(4, 2);
            this.str = str;
        }

        @Override
        public void p1() {
            System.out.println(str);
        }

        @Override
        public Rec f() {
            return new Rec(21);
        }

        @Override
        public IntegerArray fArr(IntegerArray a, IntegerRef i) {
            IntegerArray res = new IntegerArray(a.getBegin(), a.getEnd());
            for (int index = 0; index < a.size(); index++) {
                res.set(index, a.get(index) + 2);
            }
            System.out.print("Returning from Java: ");
            for (int index = 0; index < res.size(); index++) {
                System.out.format("%d, ", res.get(index));
            }
            System.out.println();
            i.setValue(i.getValue() + 2);
            return res;
        }
    }

    public static class GrandChild extends Child {

        List<Integer> vec;

        public GrandChild(List<Integer> data) {
            super(4, 2, 1);
            this.vec = data;
        }

        @Override
        public void p1() {
            for (var i : vec) {
                System.out.print(i + " ");
            }
            System.out.println();
        }

        @Override
        public void p2(Rec r, int i) {
            System.out.format("P2 from Java %d: %d\n", i, r.getI());
        }

        @Override
        public Rec f() {
            return new Rec(100);
        }
    }

    public static void testArray(Root r, IntegerArray input) {
        IntegerRef i = new IntegerRef(0);
        IntegerArray res = TestPackage.fArrDisp(r, input, i);
        System.out.print("Got in Java: ");
        for (int index = 0; index < res.size(); index++) {
            System.out.format("%d, ", res.get(index));
        }
        System.out.println();
        System.out.format("i = %d\n", i.getValue());
    }

    public static void main(String[] args) {
        Root root = new Root(1, 2);
        root.p1();

        Child child = new Child(3, 4, 5);
        child.p1();

        OtherChild other = new OtherChild("Java inherited");
        other.p1();

        GrandChild grand = new GrandChild(List.of(3, 4, 5));
        grand.p1();

        TestPackage.pRoot(root);
        TestPackage.pRoot(child);
        TestPackage.pRoot(other);
        TestPackage.pRoot(grand);

        Rec r1 = TestPackage.fRec(root);
        System.out.format("r1 = %d\n", r1.getI());
        Rec r2 = TestPackage.fRec(other);
        System.out.format("r2 = %d\n", r2.getI());
        Rec r3 = TestPackage.fRec(grand);
        System.out.format("r3 = %d\n", r3.getI());

        TestPackage.p2Child(child, r1, 1);
        TestPackage.p2Child(grand, r2, 1);

        IntegerArray arr = new IntegerArray(1, 5);
        for (int i = arr.getBegin(); i <= arr.getEnd(); i++) {
            arr.setUnslided(i, i);
        }
        testArray(root, arr);
        testArray(other, arr);
    }
}
